src/compiler/expressionToTypeNode.ts

```diff
@@ -0,0 +1,504 @@
+import {
+    AccessorDeclaration,
+    AllAccessorDeclarations,
+    ArrayLiteralExpression,
+    ArrowFunction,
+    AsExpression,
+    BinaryExpression,
+    BindingElement,
+    ClassExpression,
+    CompilerOptions,
+    Debug,
+    ElementAccessExpression,
+    ExportAssignment,
+    Expression,
+    forEachReturnStatement,
+    FunctionExpression,
+    FunctionLikeDeclaration,
+    GetAccessorDeclaration,
+    getEffectiveReturnTypeNode,
+    getEffectiveTypeAnnotationNode,
+    getJSDocTypeAssertionType,
+    getStrictOptionValue,
+    HasInferredType,
+    Identifier,
+    IntersectionTypeNode,
+    isBlock,
+    isConstTypeReference,
+    isDeclarationReadonly,
+    isEntityNameExpression,
+    isGetAccessor,
+    isIdentifier,
+    isJSDocTypeAssertion,
+    isKeyword,
+    isParameter,
+    isPrimitiveLiteralValue,
+    isShorthandPropertyAssignment,
+    isSpreadAssignment,
+    isValueSignatureDeclaration,
+    isVarConstLike,
+    JSDocSignature,
+    MethodDeclaration,
+    Node,
+    NodeArray,
+    NodeFlags,
+    nodeIsMissing,
+    ObjectLiteralExpression,
+    ParameterDeclaration,
+    ParenthesizedExpression,
+    ParenthesizedTypeNode,
+    PrefixUnaryExpression,
+    PropertyAccessExpression,
+    PropertyAssignment,
+    PropertyDeclaration,
+    PropertyName,
+    PropertySignature,
+    SetAccessorDeclaration,
+    SignatureDeclaration,
+    SymbolAccessibility,
+    SyntacticTypeNodeBuilderContext,
+    SyntaxKind,
+    TypeAssertion,
+    TypeNode,
+    TypeParameterDeclaration,
+    UnionTypeNode,
+    VariableDeclaration,
+} from "./_namespaces/ts";
+
+export function createSyntacticTypeNodeBuilder(options: CompilerOptions) {
+    const strictNullChecks = getStrictOptionValue(options, "strictNullChecks");
+
+    return {

```

Comment:
I have two questions with regard to this:
1. The true/false result is less important, the errors are generated through calling `reportInferenceFallback` on the actual node that causes the issue. Should this be caching the nodes (as there could be multiple) that caused the fallback to inference?

2. Currently we return true/false but the final end state is that this is integrated into the `NodeBuilder` in the checker and that it will actually generate new type nodes instead. I don't think caching the synthetic type nodes is something the current `NodeBuilder` does. Am I wrong about this?

Reply:
> The true/false result is less important, the errors are generated through calling reportInferenceFallback on the actual node that causes the issue. Should this be caching the nodes (as there could be multiple) that caused the fallback to inference?

You could, but probably don't need to. If you have a cached result, the diagnostics *should* have already been emitted, so you'll just issue duplicates that'll get deduped anyway.

> I don't think caching the synthetic type nodes is something the current NodeBuilder does. Am I wrong about this?

Yes, actually. `typeParameterToName` has caching, since the same parameter often gets serialized many times, though you're right that there's no top-level cache.